core/trino-main/src/main/java/io/trino/metadata/CatalogMetadata.java

```diff
@@ -36,110 +35,121 @@
+        if (catalogName.equals(informationSchemaTransaction.getCatalogName())) {
+            return informationSchemaTransaction.getConnectorMetadata(session);
         }
-        if (catalogName.equals(systemTablesId)) {
-            return systemTables.apply(session);
+        if (catalogName.equals(systemTransaction.getCatalogName())) {
+            return systemTransaction.getConnectorMetadata(session);
         }
         throw new IllegalArgumentException("Unknown connector id: " + catalogName);
     }
 
     public ConnectorTransactionHandle getTransactionHandleFor(CatalogName catalogName)
     {
-        if (catalogName.equals(this.catalogName)) {
-            return transactionHandle;
+        if (catalogName.equals(catalogTransaction.getCatalogName())) {
+            return catalogTransaction.getTransactionHandle();
         }
-        if (catalogName.equals(informationSchemaId)) {
-            return informationSchemaTransactionHandle;
+        if (catalogName.equals(informationSchemaTransaction.getCatalogName())) {
+            return informationSchemaTransaction.getTransactionHandle();
         }
-        if (catalogName.equals(systemTablesId)) {
-            return systemTablesTransactionHandle;
+        if (catalogName.equals(systemTransaction.getCatalogName())) {
+            return systemTransaction.getTransactionHandle();
         }
         throw new IllegalArgumentException("Unknown connector id: " + catalogName);
     }
 
     public CatalogName getConnectorIdForSchema(CatalogSchemaName schema)
     {
         if (schema.getSchemaName().equals(INFORMATION_SCHEMA_NAME)) {
-            return informationSchemaId;
+            return informationSchemaTransaction.getCatalogName();
         }
-        return catalogName;
+        return catalogTransaction.getCatalogName();
     }
 
     public CatalogName getConnectorId(Session session, QualifiedObjectName table)
     {
         if (table.getSchemaName().equals(INFORMATION_SCHEMA_NAME)) {
-            return informationSchemaId;
+            return informationSchemaTransaction.getCatalogName();
         }
 
-        if (systemTables.apply(session).getTableHandle(session.toConnectorSession(systemTablesId), table.asSchemaTableName()) != null) {
-            return systemTablesId;
+        if (systemTransaction.getConnectorMetadata(session).getTableHandle(session.toConnectorSession(systemTransaction.getCatalogName()), table.asSchemaTableName()) != null) {
+            return systemTransaction.getCatalogName();
         }
 
-        return catalogName;
+        return catalogTransaction.getCatalogName();
+    }
+
+    public void commit()
+    {
+        informationSchemaTransaction.commit();
+        systemTransaction.commit();
+        catalogTransaction.commit();
+    }
+
+    public void abort()
+    {
+        safeAbort(informationSchemaTransaction);
+        safeAbort(systemTransaction);
+        safeAbort(catalogTransaction);
```

Comment:
should you throw from top level `abort` (at the end) if there was an exception in one of `safeAbort` calls?

Reply:
Abort is not allowed to fail, and runs in a non-connected thread to any exception thrown is just lost.